### PR TITLE
[FEAT] Improve error message of isEnum validator

### PR DESCRIPTION
**Pull request by Bendakh**
## Description

This PR aims to improve the error message of the `IsEnum` validator. The new error will display the invalid value of the property that was not validated.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: packages/custom-decorators/test/is-enum-decorator.test.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
+import { IsEnum } from '../src/is-enum.decorator';
+import { Validator, ValidatorOptions } from 'class-validator';
+
+function validateValues(
+  object: { testProperty: any },
+  values: any[],
+  validatorOptions?: ValidatorOptions,
+): Promise<any> {
```
Comment (amarlankri) [inline packages/custom-decorators/test/is-enum-decorator.test.ts:8]: Avoid to use `any`. If there is no way to type the function properly (using an interface or generics), it is better to use `unknown`. Here, actually, it is not big deal because it is a test implementation, but it is better for readability and test maintenance.

File: packages/custom-decorators/test/is-enum-decorator.test.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+      expect(errors[0].target).toEqual(object);
+      expect(errors[0].property).toEqual('testProperty');
+      expect(errors[0].constraints).toEqual({ [validationType]: messages[messagesIncrementor] });
+      expect(errors[0].value).toEqual(value);
+      messagesIncrementor++;
+    });
+  });
+
+  return Promise.all(promises);
+}
```
Comment (amarlankri) [inline packages/custom-decorators/test/is-enum-decorator.test.ts:46]: In my opinion, those utility function adds more complexity than they solve. And both are used once, so they didn't reduce duplication.

Do you think they are useful? 🤔

File: packages/custom-decorators/src/is-enum.decorator.ts
Comment (amarlankri) [inline packages/custom-decorators/src/is-enum.decorator.ts:1]: Actually, I noticed that it is possible to use the decorator as a normal function (without annotation `@`). As a result, there is no need to duplicate the implementation of the original decorator. Instead, you can write:

```ts
import { IsEnum as OriginalIsEnum, ValidationOptions } from 'class-validator';

export const IsEnum = (entity: object, validationOptions?: ValidationOptions): PropertyDecorator => OriginalIsEnum(entity, { message: '$property has the value $value but must be one of the following values: $constraint2', ...validationOptions });
```

File: tsconfig.json [hunk 3]
```diff
@@ -25,7 +25,8 @@
 
     /* Experimental Options */                
     "experimentalDecorators": true,        /* Enables experimental support for ES7 decorators. */
-    "emitDecoratorMetadata": true          /* Enables experimental support for emitting type metadata for decorators. */
+    "emitDecoratorMetadata": true,         /* Enables experimental support for emitting type metadata for decorators. */
+    "strictPropertyInitialization": false,      /* Disables mandatory property initialization */
```
Comment (amarlankri) [inline tsconfig.json:29]: Why did you change this config?

File: packages/custom-decorators/test/is-enum-decorator.test.ts (hunk 1 above)
Comment (Bendakh) [reply packages/custom-decorators/test/is-enum-decorator.test.ts:8]: I see. I used the any because I got an error with `Object.entries` saying that it can't accept arguments of type unkown

File: tsconfig.json (hunk 3 above)
Comment (amarlankri) [reply tsconfig.json:29]: Not sure it is a good idea to "relax" a TypeScript option just for a test implementation. You cannot initialise them instead?

File: packages/custom-decorators/package.json [hunk 4]
```diff
@@ -36,7 +36,8 @@
     "url": "https://github.com/algoan/nestjs-components/issues"
   },
   "dependencies": {
-    "jwt-decode": "^4.0.0"
+    "jwt-decode": "^4.0.0",
+    "class-validator": "^0.14.0"
```
Comment (g-ongenae) [inline packages/custom-decorators/package.json:40]: Maybe this should be a peer dependency instead?

Comment (g-ongenae) [reply packages/custom-decorators/package.json:40]: Ah! I didn't know that. Then, I think it would be better to create a new package `@algoan/nestjs-class-validator` instead of updating `@algoan/nestjs-custom-decorators`.

The current decorators in this package are dedicated to parameters instead of validation and are few in numbers. Adding the whole class-validator will create some confusion, I fear. Hence, a new package.

File: tsconfig.json (hunk 3 above)
Comment (Bendakh) [reply tsconfig.json:29]: I'll do that instead then

File: packages/custom-decorators/package.json
Comment (g-ongenae) [inline packages/custom-decorators/package.json:1]: This file should not be changed by this PR anymore.

File: tsconfig.json
Comment (g-ongenae) [inline tsconfig.json:1]: Same.

File: packages/nestjs-class-validators/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,2 @@
+export { IsEnum } from './is-enum.decorator';
+export * from 'class-validator';
```
Comment (g-ongenae) [inline packages/nestjs-class-validators/src/index.ts:2]: Shouldn't it export everything except the original `IsEnum`?

File: packages/nestjs-class-validators/package.json [hunk 6]
```diff
@@ -0,0 +1,14 @@
+{
+  "name": "nestjs-class-validators",
```
Comment (g-ongenae) [inline packages/nestjs-class-validators/package.json:2]:
```suggestion
  "name": "@algoan/nestjs-class-validators",
```

It also should be added it here:
https://github.com/algoan/nestjs-components/blob/5e327217a30fe766eb12b214620a9bd19aa14974/package.json#L91-L92

File: packages/nestjs-class-validators/package.json
Comment (g-ongenae) [inline packages/nestjs-class-validators/package.json:1]: The folder can be named `class-validators` directly for coherency with the other packages. (I.e. nestjs in the package name, not in the folder name.)

File: packages/custom-decorators/test/is-enum-decorator.test.ts [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+import { Validator, ValidatorOptions } from 'class-validator';
+
+function validateValues(
+  object: { testProperty: any },
+  values: any[],
+  validatorOptions?: ValidatorOptions,
+): Promise<any> {
+  const validator = new Validator();
+  const promises = values.map((value) => {
+    object.testProperty = value;
```
Comment (qhdinh) [inline packages/custom-decorators/test/is-enum-decorator.test.ts]: there is just one instance of object but you set and test many async validation parallelly on the same property, I'm not sure it works correctly as expected

File: packages/custom-decorators/src/is-enum.decorator.ts [hunk 8]
```diff
@@ -0,0 +1,7 @@
+import { IsEnum as OriginalIsEnum, ValidationOptions } from 'class-validator';
+
+export const IsEnum = (entity: object, validationOptions?: ValidationOptions): PropertyDecorator =>
+  OriginalIsEnum(entity, {
+    message: '$property has the value $value but must be one of the following values: $constraint2',
```
[the remark this answers is not in the document]
Comment (g-ongenae) [reply packages/custom-decorators/src/is-enum.decorator.ts]: To the `validEnumEntities`, see [code](https://github.com/typestack/class-validator/blob/5ea098fb988ce4e316c1b3bad496f6db80c44d81/src/decorator/typechecker/IsEnum.ts#L30).

[the remark this answers is not in the document]
Comment (qhdinh) [reply packages/custom-decorators/src/is-enum.decorator.ts]: Okay, thanks. I don't see they indicate in the offical doc page

File: packages/custom-decorators/test/is-enum-decorator.test.ts [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    return validator.validate(object, validatorOptions).then((errors) => {
+      expect(errors.length).toEqual(0);
+      if (errors.length > 0) {
+        console.log(`Unexpected errors: ${JSON.stringify(errors)}`);
+        throw new Error('Unexpected validation errors');
+      }
+    });
```
[the remark this answers is not in the document]
Comment (qhdinh) [reply packages/custom-decorators/test/is-enum-decorator.test.ts]: But in a test there must be expecting something. In your test, you just call `validateValues()`. 
I think it'd be cleaner `validateValues()` just return a list of validation errors. And in the test you expect the list after the function call.

File: packages/nestjs-class-validators/package.json
Comment (qhdinh) [reply packages/nestjs-class-validators/package.json:1]: This should be put in our mono repo instead 🤪

File: packages/custom-decorators/test/is-enum-decorator.test.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (Bendakh) [reply packages/custom-decorators/test/is-enum-decorator.test.ts]: I see your point. I didn't change much of the implementation for the tests to as I said in another comment avoid breaking anything. I will update this

File: packages/class-validators/src/is-enum.decorator.ts [hunk 10]
```diff
@@ -0,0 +1,7 @@
+import { IsEnum as OriginalIsEnum, ValidationOptions } from 'class-validator';
+
+export const IsEnum = (entity: object, validationOptions?: ValidationOptions): PropertyDecorator =>
```
Comment (amarlankri) [inline packages/class-validators/src/is-enum.decorator.ts:3]: Since this function is part of the public API, you should probably add JSDoc.

**Review by amarlankri — APPROVED** [review]

**Review by qhdinh — APPROVED** [review]

**Review by g-ongenae — APPROVED** [review]
